src/runtime/client-event-manager.js

```diff
@@ -15,33 +15,133 @@
+}
+
+/**
+ * Throws an error if the event is invalid.
+ * @param {!../api/client-event-manager-api.ClientEvent} event
+ */
+function validateEvent(event) {
+  if (!isObject(event)) {
+    throw new Error('Event must be a valid object');
+  }
+
+  if (!isEnumValue(AnalyticsEvent, event.eventType)) {
+    throw new Error(createEventErrorMessage('eventType', event.eventType));
+  }
+
+  if (!isEnumValue(EventOriginator, event.eventOriginator)) {
+    throw new Error(createEventErrorMessage('eventOriginator',
+        event.eventOriginator));
+  }
+
+  if (!isObject(event.additionalParameters)
+      && event.additionalParameters !== null) {
+    if (event.additionalParameters !== undefined) {
+      throw new Error(createEventErrorMessage('additionalParameters',
+          event.additionalParameters));
+    }
+    event.additionalParameters = null;
+  }
+
+
+  if (event.isFromUserAction !== null && !isBoolean(event.isFromUserAction)) {
+    throw new Error(createEventErrorMessage('isFromUserAction',
+        event.isFromUserAction));
+  }
+}
+
+/** @implements {../api/client-event-manager-api.ClientEventManagerApi} */
 export class ClientEventManager {
   constructor() {
+    /** @private {!Array<function(!../api/client-event-manager-api.ClientEvent)>} */
+    this.listeners_ = [];
+
+    /** @private {!Array<function(!../api/client-event-manager-api.ClientEvent):!FilterResult>} */
+    this.filterers_ = [];
+
+    /** @private {?Promise} */
+    this.lastAction_ = null;
   }
 
   /**
    * @overrides
    */
-  registerEventListener(unusedCallback) {
+  registerEventListener(listener) {
+    if (!isFunction(listener)) {
+      throw new Error('Event manager listeners must be a function');
+    }
+    this.listeners_.push(listener);
   }
 
   /**
    * @overrides
    */
-  registerEventFilterer(unusedCallback) {
+  registerEventFilterer(filterer) {
+    if (!isFunction(filterer)) {
+      throw new Error('Event manager filterers must be a function');
+    }
+    this.filterers_.push(filterer);
   }
 
   /**
    * @overrides
    */
   logEvent(event) {
-    //TODO(mborof): the API must be used somewhere or presubmit fails.
-    //              Remove this line once all code is implemented
-    if (event === FilterResult.CANCEL_EVENT) {
-      return;
-    }
-    return;
+    validateEvent(event);
+    //this will either resolve once the event is filtered out or once every
+    //listener is called.  This only guarantees the listener returned, it
+    //does not guarantee the listener finished anything it might have done
+    //asynchronously.
+    this.lastAction_ = new Promise(resolve => {
+      for (let filterer = 0; filterer < this.filterers_.length; filterer++) {
+        if (this.filterers_[filterer](event) === FilterResult.CANCEL_EVENT) {
+          resolve();
+          return;
+        }
+      }
+
+      const promises = [];
+      let listenerNum = 0;
+      //builds an array of 1 promise per listener
```

Comment:
since this whole part is running asynchornously after validate event returns to the caller, any reason you need to chain promises?

sounds like what you want to do is dispatch events to its listeners in an event queue pattern.

I'd like to understand the need for doing this though. If the entire filtering and invoking listeners is happening asynchronously as long as the event is validated, nothing is blocking on it. 

Another way to make it asynchronous is create a private method that returns the result of filtering as a promise. Since processing is blocked on filtering, you can make dispatching an event block on filter result. Example
/**
 * @param {!ClientEvent} event
 * @return {!Promise<boolean>}
 */
filterEvent(event) {
  let shouldProcess = true;
  this.filterers_.foreach(filterer => {
    // If any fliterer wants to cancel this event, cancel it
    if (filterer(event) == FilterResult.CANCEL_EVENT) {
      shouldProcess = false;
      break;
    }
  });
  return Promise.resolve(shouldProcess);
}
/**
 * @param {ClientEvent} event
 */
dispatchEvent(event) {
  this.listeners_.foreach(listener => {
    setTimeout(() => listener(event));
  });
}

logEvent(event) {
  return validateEvent(event).then(() => {
    // Filtering and dispatching the event are asynchronous
    return filterEvent(event);
  }).then(process => {
    if (process) {
      dispatchEvent(event);
    }
  });

Reply:
We discussed and decided that 1 promise would be fine.  There is no need to asynchronously call each individual listener and then know that each of those listeners was called.

Reply:
To conclude, since the listeners are all internal callbacks (propensity service, analytics service and AMP), we can assure the quality of these listeners and invoke them synchronously.